tor module: add support for transparent proxy and dns

offlinehacker opened the pull request:
This pull request adds support for tor transparent proxy


<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

offlinehacker committed:
nixos/tor: add support for transparent proxy and dns

oxij commented:
A quote from SLNOS discussion:

> * `AutomapHostsSuffixes` must either be made configurable or be
> removed. The patch fixes the value to the default value. Makes no
> sense.
>
> I also don't want to have `.exit` in there for security reasons.
>
> * Same for `ClientRejectInternalAddresses` and
> `ClientDNSRejectInternalAddresses`.
>
> * `ClientRejectInternalAddresses` if made configurable should be moved
> under `optionalString cfg.client.transparentProxy.enable` block. It
> configures the proxy not the DNS.

offlinehacker commented:
So i removed `ClientRejectInternalAddresses` and `ClientDNSRejectInternalAddresses` as these are set by default and i added option for `AutomapHostsSuffixes`.

oxij commented:
> So i removed `ClientRejectInternalAddresses` and `ClientDNSRejectInternalAddresses` as these are set by default and i added option for `AutomapHostsSuffixes`.

* Please set all new `enable` to `false` by default.
* Also set `automapHostsSuffixes` to `[".onion" ".exit"]` by default (it's
  tor's default).

Some more personal nitpicks:

* My personal OCD requires all options of newly added `mkOption`s to be
  consistently ordered (e.g. `type`, followed `default`, `example`,
  `description`, any consistent order is fine).

* If I were you I would also generalize other places that use isolation
  (both `SOCKSPort`s) with your new `isolationOptions` (just make it
  into a function that takes its `default` as an argument). It can be
  done later, though.

offlinehacker commented:
Ok, fixed, is this ok now?

oxij commented:
LGTM. I haven't tested this, though.
